core/src/main/java/com/google/cloud/sql/core/AccessTokenSupplier.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import com.google.auth.oauth2.AccessToken;
+import java.io.IOException;
+import java.util.Optional;
+
+/** Supplies an AccessToken to use when authenticating with the Google API. */
+@FunctionalInterface
+interface AccessTokenSupplier {
```

Comment:
Could we just use `Supplier<AccessToken>` instead here?

Reply:
No, because we `Supplier<AccessToken>` doesn't throw an exception. We need to distinguish between when `get()` throws an IOException  meaning the refresh attempt failed to make a successful request, from when it throws an unchecked exception like IllegalStateException, meaning that the token can't be refreshed due to business logic.